File: java/PrimitiveTypes.java
/**
 * Repertório de Tipos Primitivos em JAVA
 * -----------------------------------------
 * Java possui 8 tipos primitivos, divididos em:
 * - Numéricos inteiros: byte, short, int, long
 * - Numéricos de ponto flutuante: float, double
 * - Caractere: char
 * - Lógico: boolean
 *
 * Para compilar e executar:
 *   javac PrimitiveTypes.java
 *   java PrimitiveTypes
 */
public class PrimitiveTypes {

    public static void main(String[] args) {

        // ---------- Tipos Inteiros ----------
        byte tipoByte = 127;                     // 8 bits  (-128 a 127)
        short tipoShort = 32000;                 // 16 bits (-32.768 a 32.767)
        int tipoInt = 2_147_483_000;              // 32 bits (~ -2,1bi a 2,1bi)
        long tipoLong = 9_223_372_036_854_775L;   // 64 bits

        // ---------- Tipos de Ponto Flutuante ----------
        float tipoFloat = 3.14159f;               // 32 bits, precisão simples
        double tipoDouble = 3.141592653589793;    // 64 bits, precisão dupla

        // ---------- Caractere ----------
        char tipoChar = 'A';                      // 16 bits, um único caractere Unicode

        // ---------- Lógico ----------
        boolean tipoBoolean = true;               // true ou false

        System.out.println("===== REPERTÓRIO DE TIPOS PRIMITIVOS - JAVA =====\n");

        imprimirTipo("byte", tipoByte, Byte.SIZE, Byte.MIN_VALUE, Byte.MAX_VALUE);
        imprimirTipo("short", tipoShort, Short.SIZE, Short.MIN_VALUE, Short.MAX_VALUE);
        imprimirTipo("int", tipoInt, Integer.SIZE, Integer.MIN_VALUE, Integer.MAX_VALUE);
        imprimirTipo("long", tipoLong, Long.SIZE, Long.MIN_VALUE, Long.MAX_VALUE);

        System.out.println("-------------------------------------------------");
        System.out.println("float    | valor: " + tipoFloat + " | bits: 32 | ~7 dígitos de precisão");
        System.out.println("double   | valor: " + tipoDouble + " | bits: 64 | ~15 dígitos de precisão");
        System.out.println("-------------------------------------------------");
        System.out.println("char     | valor: '" + tipoChar + "' | bits: 16 | código Unicode: " + (int) tipoChar);
        System.out.println("-------------------------------------------------");
        System.out.println("boolean  | valor: " + tipoBoolean + " | valores possíveis: true / false");
        System.out.println("-------------------------------------------------\n");

        // Demonstração de conversão implícita (widening) e explícita (narrowing)
        System.out.println("===== CONVERSÕES ENTRE TIPOS =====");
        int numeroInt = 100;
        double numeroDouble = numeroInt; // widening (implícita)
        System.out.println("int -> double (implícita): " + numeroInt + " -> " + numeroDouble);

        double valorAlto = 9.99;
        int valorTruncado = (int) valorAlto; // narrowing (explícita)
        System.out.println("double -> int (explícita, com cast): " + valorAlto + " -> " + valorTruncado);
    }

    // Método auxiliar para tipos inteiros, mostrando tamanho e faixa de valores
    private static void imprimirTipo(String nome, long valor, int bits, long min, long max) {
        System.out.printf("%-8s | valor: %-20d | bits: %-3d | faixa: %d a %d%n",
                nome, valor, bits, min, max);
    }
}

File: javascript/primitiveTypes.js
/* * Repertório de Tipos Primitivos em JAVASCRIPT
 * -----------------------------------------------
 * JavaScript possui 7 tipos primitivos (ECMAScript atual):
 *   - string
 *   - number
 *   - bigint
 *   - boolean
 *   - undefined
 *   - null
 *   - symbol
 *
 * Diferente de Java, JS não distingue int/float/double: existe
 * apenas "number" (ponto flutuante de 64 bits, IEEE 754).
 *
 * Para executar:
 *   node primitiveTypes.js
 */

console.log("===== REPERTÓRIO DE TIPOS PRIMITIVOS - JAVASCRIPT =====\n");

// ---------- string ----------
const tipoString = "Olá, mundo!";
console.log(`string    | valor: "${tipoString}" | typeof: ${typeof tipoString}`);

// ---------- number ----------
const tipoNumberInt = 42;
const tipoNumberFloat = 3.14159;
console.log(`number    | valor: ${tipoNumberInt} | typeof: ${typeof tipoNumberInt} (inteiro)`);
console.log(`number    | valor: ${tipoNumberFloat} | typeof: ${typeof tipoNumberFloat} (decimal)`);
console.log(`number    | Number.MAX_SAFE_INTEGER: ${Number.MAX_SAFE_INTEGER}`);

// ---------- bigint ----------
const tipoBigInt = 9007199254740993n; // note o sufixo "n"
console.log(`bigint    | valor: ${tipoBigInt} | typeof: ${typeof tipoBigInt}`);

// ---------- boolean ----------
const tipoBoolean = true;
console.log(`boolean   | valor: ${tipoBoolean} | typeof: ${typeof tipoBoolean}`);

// ---------- undefined ----------
let tipoUndefined;
console.log(`undefined | valor: ${tipoUndefined} | typeof: ${typeof tipoUndefined}`);

// ---------- null ----------
const tipoNull = null;
console.log(`null      | valor: ${tipoNull} | typeof: ${typeof tipoNull} (curiosidade: bug histórico do JS)`);

// ---------- symbol ----------
const tipoSymbol = Symbol("identificador único");
console.log(`symbol    | valor: ${tipoSymbol.toString()} | typeof: ${typeof tipoSymbol}`);

console.log("\n===== CONVERSÕES ENTRE TIPOS (coerção) =====");

// Conversão implícita
const somaComString = "5" + 3;        // "53" (concatenação)
const somaComNumero = "5" - 3;        // 2 (coerção para número)
console.log(`"5" + 3  = ${somaComString}  (typeof: ${typeof somaComString})`);
console.log(`"5" - 3  = ${somaComNumero}  (typeof: ${typeof somaComNumero})`);

// Conversão explícita
const textoParaNumero = Number("123.45");
const numeroParaTexto = String(123.45);
const numeroParaBoolean = Boolean(0); // false
console.log(`Number("123.45") = ${textoParaNumero} (${typeof textoParaNumero})`);
console.log(`String(123.45)   = "${numeroParaTexto}" (${typeof numeroParaTexto})`);
console.log(`Boolean(0)       = ${numeroParaBoolean} (${typeof numeroParaBoolean})`);
